### PR TITLE
need marshal for proxy set trap calling arguments

### PR DESCRIPTION
**Pull request by leirocks**

### CODE REVIEW

**Comment by kunalspathak** [comment]
:shipit:

**Comment by akroshg** [comment]
>         varArgs[3] = descVar;

do we need to marshal this?

---
Refers to: lib/Runtime/Library/JavascriptProxy.cpp:1615 in d5a5e63. [](commit_id = d5a5e63d534f91417a7911f0cde8522a32f61d43, deletion_comment = False)

**Comment by akroshg** [comment]
>         varArgs[2] = newPrototype;

what about this case?

---
Refers to: lib/Runtime/Library/JavascriptProxy.cpp:1452 in d5a5e63. [](commit_id = d5a5e63d534f91417a7911f0cde8522a32f61d43, deletion_comment = False)

**Comment by kunalspathak** [comment]
>         varArgs[2] = newPrototype;

Thats true. I think instead of spot fixing we should revisit all the proxy APIs and add marshalling wherever needed for objects that are passed as parameter to the trap.

---
In reply to: [250006065](https://github.com/Microsoft/ChakraCore/pull/1648#issuecomment-250006065) [](ancestors = 250006065)

---
Refers to: lib/Runtime/Library/JavascriptProxy.cpp:1452 in d5a5e63. [](commit_id = d5a5e63d534f91417a7911f0cde8522a32f61d43, deletion_comment = False)

**Comment by akroshg** [comment]
>         varArgs[2] = newPrototype;

It would be good to have test where input is from cross context.

---
In reply to: [250006654](https://github.com/Microsoft/ChakraCore/pull/1648#issuecomment-250006654) [](ancestors = 250006654,250006065)

---
Refers to: lib/Runtime/Library/JavascriptProxy.cpp:1452 in d5a5e63. [](commit_id = d5a5e63d534f91417a7911f0cde8522a32f61d43, deletion_comment = False)

**Comment by curtisman** [comment]
>         setPropertyResult = setMethod->CallFunction(arguments);

Shouldn't the setMethod already have a cross site thunk already so you don't have to manully marshal var here?

---
Refers to: lib/Runtime/Library/JavascriptProxy.cpp:1747 in d5a5e63. [](commit_id = d5a5e63d534f91417a7911f0cde8522a32f61d43, deletion_comment = False)

**Comment by curtisman** [comment]
>         varArgs[2] = newPrototype;

All the trap function that we are about to call should have the cross site thunk if it is from another script context.  
That should be the bug if it is not, instead of manually marshalling the args.

---
In reply to: [250007075](https://github.com/Microsoft/ChakraCore/pull/1648#issuecomment-250007075) [](ancestors = 250007075,250006654,250006065)

---
Refers to: lib/Runtime/Library/JavascriptProxy.cpp:1452 in d5a5e63. [](commit_id = d5a5e63d534f91417a7911f0cde8522a32f61d43, deletion_comment = False)

**Comment by leirocks** [comment]
>         setPropertyResult = setMethod->CallFunction(arguments);

fixed, can you take a look again?

---
In reply to: [250065291](https://github.com/Microsoft/ChakraCore/pull/1648#issuecomment-250065291) [](ancestors = 250065291)

---
Refers to: lib/Runtime/Library/JavascriptProxy.cpp:1747 in d5a5e63. [](commit_id = d5a5e63d534f91417a7911f0cde8522a32f61d43, deletion_comment = False)

File: lib/Runtime/Library/JavascriptProxy.cpp [hunk 1]
```diff
@@ -1809,7 +1809,7 @@ namespace Js
         //  3. If func is either undefined or null, return undefined.
         //  4. If IsCallable(func) is false, throw a TypeError exception.
         //  5. Return func.
-        BOOL result = JavascriptOperators::GetPropertyReference(handler, methodId, &varMethod, requestContext);
+        BOOL result = JavascriptOperators::GetPropertyReference(handler, methodId, &varMethod, GetScriptContext());
```
Comment (akroshg) [inline lib/Runtime/Library/JavascriptProxy.cpp]: I think the best thing to do is leave this as requestcontext and below have call to marshal it.

---
In reply to: [81822215](https://github.com/Microsoft/ChakraCore/pull/1648#discussion_r81822215) [](ancestors = 81822215)

Comment (leirocks) [inline lib/Runtime/Library/JavascriptProxy.cpp]: done

**Comment by leirocks** [comment]
@dotnet-bot test Ubuntu ubuntu_linux_debug  please

**Comment by akroshg** [comment]
:shipit:

**Pull request merged**